relay-server/src/middlewares/body_timing.rs

```diff
@@ -0,0 +1,288 @@
+use crate::statsd::RelayTimers;
+use axum::body::{Body, HttpBody};
+use axum::http::Request;
+use hyper::body::Frame;
+use relay_statsd::metric;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use tokio::time::Instant;
+use tower::{Layer, Service};
+
+/// Middleware layer that wraps the request [`Body`] to measure body reading duration.
+///
+/// Metrics are tagged with a completion status to track whether the body:
+/// - was read to completion ("completed")
+/// - failed during reading ("failed")
+/// - was dropped before completion ("dropped")
+///
+/// The reported size reflects the actual number of bytes read, which may differ from
+/// the `Content-Length` header. For bodies that fail or are dropped mid-reading,
+/// the size represents only the bytes successfully read before the interruption.
+#[derive(Clone, Debug)]
+pub struct BodyTimingLayer;
+
+impl BodyTimingLayer {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl<S> Layer<S> for BodyTimingLayer {
+    type Service = BodyTiming<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        BodyTiming::new(inner)
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct BodyTiming<S> {
+    inner: S,
+}
+
+impl<S> BodyTiming<S> {
+    pub fn new(inner: S) -> Self {
+        Self { inner }
+    }
+}
+
+impl<S> Service<Request<Body>> for BodyTiming<S>
+where
+    S: Service<Request<Body>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        let (parts, body) = req.into_parts();
+        let timed_body = Body::new(TimedBody::new(body, parts.uri.to_string()));
+        let request = Request::from_parts(parts, timed_body);
+        self.inner.call(request)
+    }
+}
+
+#[derive(Debug)]
+struct TimedBody {
+    inner: Body,
+    reading_started_at: Option<Instant>,
+    size: u32,
+    route: String,
+}
+
+impl TimedBody {
+    fn new<T>(inner: Body, route: T) -> Self
```

Comment:
Since the route is always converted into a String, I'd just require a `String` here, this makes the allocation explicit to the caller and less error prone (maybe the caller already had a String but passed a reference by mistake).